ksonnet-gen/ksonnet/emit.go

```diff
@@ -804,44 +808,79 @@ func (p *property) emitHelper(
 		} else {
 			body = fmt.Sprintf("%s({%s: %s})", *parentMixinName, fieldName, paramName)
 		}
-		line := fmt.Sprintf("%s %s,", signature, body)
+		line := fmt.Sprintf("%s %s,", setterSignature, body)
 		m.writeLine(line)
 	} else if p.schemaType != nil {
 		paramType := *p.schemaType
 
-		var body string
+		//
+		// Generate both setter and mixin functions for some property. For
+		// example, we emit both `metadata.setAnnotations({foo: "bar"})`
+		// (which replaces a set of annotations with given object) and
+		// `metadata.mixinAnnotations({foo: "bar"})` (which replaces only
+		// the `foo` key, if it exists.)
+		//
+
+		var setterBody string
+		var mixinBody string
+		emitMixin := false
 		switch paramType {
 		case "array":
 			if parentMixinName == nil {
-				body = fmt.Sprintf(
+				setterBody = fmt.Sprintf(
+					"if std.type(%s) == \"array\" then {%s: %s} else {%s: [%s]}",
+					paramName, fieldName, paramName, fieldName, paramName,
+				)
+				mixinBody = fmt.Sprintf(
 					"if std.type(%s) == \"array\" then {%s+: %s} else {%s+: [%s]}",
 					paramName, fieldName, paramName, fieldName, paramName,
 				)
+				emitMixin = true
 			} else {
-				body = fmt.Sprintf(
+				setterBody = fmt.Sprintf(
+					"if std.type(%s) == \"array\" then %s({%s: %s}) else %s({%s: [%s]})",
+					paramName, *parentMixinName, fieldName, paramName, *parentMixinName,
+					fieldName, paramName,
+				)
+				mixinBody = fmt.Sprintf(
 					"if std.type(%s) == \"array\" then %s({%s+: %s}) else %s({%s+: [%s]})",
 					paramName, *parentMixinName, fieldName, paramName, *parentMixinName,
 					fieldName, paramName,
 				)
+				emitMixin = true
 			}
 		case "integer", "string", "boolean":
 			if parentMixinName == nil {
-				body = fmt.Sprintf("{%s: %s}", fieldName, paramName)
+				setterBody = fmt.Sprintf("{%s: %s}", fieldName, paramName)
 			} else {
-				body = fmt.Sprintf("%s({%s: %s})", *parentMixinName, fieldName, paramName)
+				setterBody = fmt.Sprintf("%s({%s: %s})", *parentMixinName, fieldName, paramName)
 			}
 		case "object":
 			if parentMixinName == nil {
-				body = fmt.Sprintf("{%s+: %s}", fieldName, paramName)
+				setterBody = fmt.Sprintf("{%s: %s}", fieldName, paramName)
+				mixinBody = fmt.Sprintf("{%s+: %s}", fieldName, paramName)
+				emitMixin = true
```

Comment:
`emitMixin` is set to `true` in both cases, can you take it out of the conditional? Same with above line 838.